src/test/scala/uk/gov/nationalarchives/tdr/api/routes/FileMetadataRouteSpec.scala

```diff
@@ -0,0 +1,157 @@
+    checkFileMetadataExists(response.data.get.addFileMetadata(1).fileId)
+
+  }
+
+  "addFileMetadata" should "not allow updating of file metadata with incorrect authorisation" in {
+    val response: GraphqlMutationData = runTestMutation("mutation_alldata", invalidBackendChecksToken())
+
+    response.errors should have size 1
+    response.errors.head.extensions.get.code should equal("NOT_AUTHORISED")
+    checkNoFileMetadataAdded()
+  }
+
+  "addFileMetadata" should "not allow updating of file metadata with incorrect client role" in {
+    val response: GraphqlMutationData = runTestMutation("mutation_alldata", validBackendChecksToken("antivirus"))
+
+    response.errors should have size 1
+    response.errors.head.extensions.get.code should equal("NOT_AUTHORISED")
+    checkNoFileMetadataAdded()
+  }
+
+  "addFileMetadata" should "throw an error if the field file property name is not provided" in {
+    val expectedResponse: GraphqlMutationData = expectedMutationResponse("data_fileproperty_missing")
+    val response: GraphqlMutationData = runTestMutation("mutation_missingfileproperty", validBackendChecksToken("checksum"))
+
+    response.errors.head.message should equal (expectedResponse.errors.head.message)
+    checkNoFileMetadataAdded()
+  }
+
+  "addFileMetadata" should "throw an error if the field file id is not provided" in {
+    val expectedResponse: GraphqlMutationData = expectedMutationResponse("data_fileid_missing")
+    val response: GraphqlMutationData = runTestMutation("mutation_missingfileid", validBackendChecksToken("checksum"))
+
+    response.errors.head.message should equal (expectedResponse.errors.head.message)
+    checkNoFileMetadataAdded()
+  }
+
+  "addFileMetadata" should "throw an error if the value is not provided" in {
+    val expectedResponse: GraphqlMutationData = expectedMutationResponse("data_value_missing")
+    val response: GraphqlMutationData = runTestMutation("mutation_missingvalue", validBackendChecksToken("checksum"))
+
+    response.errors.head.message should equal (expectedResponse.errors.head.message)
+    checkNoFileMetadataAdded()
+  }
+
+  "addFileMetadata" should "throw an error if the file id does not exist" in {
+    val expectedResponse: GraphqlMutationData = expectedMutationResponse("data_fileid_not_exists")
+    val response: GraphqlMutationData = runTestMutation("mutation_fileidnotexists", validBackendChecksToken("checksum"))
+
+    response.errors.head.message should equal (expectedResponse.errors.head.message)
+    checkNoFileMetadataAdded()
+  }
+
+  "addFileMetadata" should "throw an error if one file id does not exist and one does" in {
+    val expectedResponse: GraphqlMutationData = expectedMutationResponse("data_one_fileid_not_exists")
+    val response: GraphqlMutationData = runTestMutation("mutation_onefileidnotexists", validBackendChecksToken("checksum"))
+
+    response.errors.head.message should equal (expectedResponse.errors.head.message)
+    checkNoFileMetadataAdded()
```

Comment:
Yeah, that's the idea. If the lambda fails, we know that all file updates passed to it failed instead of having to parse through the logs trying to work out which one failed and which one succeeded.

Reply:
Also, I'm changing it to remove the validation tags and just do SQL inserts and I checked and postgres won't insert any rows if one of the rows has an integrity error.

Reply:
Whoops, I meant to write "just thinking" earlier. "Just think" sounds quite patronising, sorry. 🤦‍♀️ 

Does that mean that a single invalid update can accidentally break some valid updates, since they'll go back onto the SQS queue at the same time and probably get picked up by the same Lambda execution?

Reply:
It does but as the management of messages on the queue is handled by AWS and lambas can only succeed or fail, the alternative is ignoring the failure, the lambda succeeds and the failure is lost. 
If we leave it as is and let it fall into the dead letter queue, we can process it from there.
Given that it's only automated processes calling this code, the odds of getting an invalid id are pretty low anyway. If it does turn out to be a problem, we can tweak the batch size to minimise the damage

Reply:
I agree that failures should go to the dead letter queue. But how do we avoid messages about valid files ending up in the dead letter queue just because they happened to be in the same batch?

Annoying concurrency issues will happen at some point, and they'll be a pain to debug. Easier to fix it now when we're actively working on this part of the system.

Reply:
We had a chat about this last week, and decided:

- The lambda will still process files in batches
- It will send updates to the API individually
- The lambda will delete messages from the queue for every file update that succeeds

This means that failed messages will get retried up to the limit, and will then be sent to the dead letter queue.